src/puppetlabs/trapperkeeper/bootstrap.clj

```diff
@@ -132,30 +135,168 @@
+        (throw (bootstrap-error entry bootstrap-file line-number message))))))
+
+(schema/defn remove-duplicate-entries :- [AnnotatedBootstrapEntry]
+  "Removes any duplicate entries from the list of AnnotatedBootstrapEntry maps.
+   A entry is considered a duplicate if the :entry key is the same between two
+   entries.
+   Logs warnings for each duplicate found."
+  [entries :- [AnnotatedBootstrapEntry]]
+  (let [into-ordered-list
+        ; Accumulates bootstrap entries into :set to test for duplicates, and
+        ; accumulates the AnnotatedBootstrapEntry's into :ordered-entries so
+        ; that the order of their line numbers is maintained
+        (fn [acc annotated-entry]
+          (if (contains? (:set acc) (:entry annotated-entry))
+            (do (log/warn
+                 (format "Duplicate bootstrap entry found for service '%s' on line number '%s' in file '%s'"
+                         (:entry annotated-entry)
+                         (:line-number annotated-entry)
+                         (:bootstrap-file annotated-entry)))
+                acc)
+            {:set (conj (:set acc) (:entry annotated-entry))
+             :ordered-entries (conj (:ordered-entries acc) annotated-entry)}))]
+    ; Reduce the bootstrap entries into an ordered list of unique entries and return it
+    (:ordered-entries (reduce
+                       into-ordered-list
+                       {:set #{} :ordered-entries []}
+                       entries))))
 
-   In the case of an empty file, nil would be returned by line-seq, so we remove
-   all the nils at the end"
-  [files :- [(schema/protocol io/IOFactory)]]
-  (remove nil? (flatten (map #(line-seq (io/reader %)) files))))
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Public
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
 (schema/defn parse-bootstrap-configs! :- [(schema/protocol services/ServiceDefinition)]
+  [configs :- [schema/Str]]
   "Parse multiple trapperkeeper bootstrap configuration files and return the
-  service graph that is the result of merging the graphs of all of the services
-  specified in the configuration files."
-  [configs :- [(schema/protocol io/IOFactory)]]
-  (let [lines (chain-files configs)]
-    (when (empty? lines) (throw (Exception. "Empty bootstrap config file")))
-    (for [line (map remove-comments lines)
-          :when (not (empty? line))]
-      (let [{:keys [namespace service-name]} (parse-bootstrap-line! line)]
-        (resolve-service! namespace service-name)))))
+  service graph that is the result of merging the graphs of all of the
+  services specified in the configuration files."
+  ; We remove the duplicate entries to allow the user to have duplicate entries in their
+  ; bootstrap files. If we didn't remove them, it would look like two services were trying
+  ; to implement the same protocol when we check for duplicate service implementations
```

Comment:
Ah, I see. I was reading "duplicate entries" as meaning "duplicate protocol implementations" (I think because that's what "duplicate <some other words I wasn't paying full attention to>" means everywhere else). This makes much more sense. Maybe we could use a term other than "duplicate entries" to make this a little clear? IDK what though. Feel free to ignore this because obviously I can't read.